### PR TITLE
Add Firewall Config resource

### PR DESCRIPTION
**Pull request by sueplex**
Adds a new resource and client connector for the Firewall configuration API

Supports updating custom rules, ip blocking rules and managed rulesets

### CODE REVIEW

File: vercel/provider.go [hunk 1]
```diff
@@ -57,6 +57,7 @@ func (p *vercelProvider) Resources(_ context.Context) []func() resource.Resource
 		newEdgeConfigResource,
 		newEdgeConfigSchemaResource,
 		newEdgeConfigTokenResource,
+		newFirewallConfigResource,
```
Comment (dglsparsons) [inline vercel/provider.go:60]: is there value in having a DataSource too? (probably not?)

File: vercel/resource_attack_challenge_mode.go [hunk 2]
```diff
@@ -63,7 +63,7 @@ Attack Challenge Mode prevent malicious traffic by showing a verification challe
 				PlanModifiers: []planmodifier.String{stringplanmodifier.UseStateForUnknown()},
 			},
 			"project_id": schema.StringAttribute{
-				Description:   "The ID of the Project to adjust the CPU for.",
+				Description:   "The ID of the Project to toggle Attack Challenge Mode on.",
```
Comment (dglsparsons) [inline vercel/resource_attack_challenge_mode.go:66]: good spot.

File: vercel/resource_firewall_config.go [hunk 3]
```diff
@@ -0,0 +1,878 @@
...
+				Description: "The managed rulesets that are enabled.",
+				Blocks: map[string]schema.Block{
+					"owasp": schema.SingleNestedBlock{
+						Attributes: map[string]schema.Attribute{
+							"xss": schema.SingleNestedAttribute{
+								Optional: true,
+								Attributes: map[string]schema.Attribute{
+									"active": schema.BoolAttribute{
+										Optional: true,
+									},
```
Comment (dglsparsons) [inline vercel/resource_firewall_config.go:52]: Descriptions missing on the fields.

File: vercel/resource_firewall_config.go [hunk 4]
```diff
@@ -0,0 +1,878 @@
...
+type Redirect struct {
+	Location  types.String `tfsdk:"location"`
+	Permanent types.Bool   `tfsdk:"permanent"`
+}
+
+type RateLimit struct {
+	Algo   types.String   `tfsdk:"algo"`
+	Window types.Int64    `tfsdk:"window"`
+	Limit  types.Int64    `tfsdk:"limit"`
+	Keys   []types.String `tfsdk:"keys"`
```
Comment (dglsparsons) [inline vercel/resource_firewall_config.go:577]: types.List might be nicer 🤷

File: vercel/resource_firewall_config.go [hunk 5]
```diff
@@ -0,0 +1,878 @@
...
+		r.Active = ref.Active
+	}
+
+	return r
+}
+
+type Mitigate struct {
+	Action         types.String `tfsdk:"action"`
+	RateLimit      *RateLimit   `tfsdk:"rate_limit"`
+	Redirect       *Redirect    `tfsdk:"redirect"`
```
Comment (dglsparsons) [inline vercel/resource_firewall_config.go:532]: Could consider using `types.Object` rather than `*Redirect` for these kinda fields. Not a biggie if it makes the code a lot messier though.

File: vercel/provider.go (hunk 1 above)
Comment (sueplex) [reply vercel/provider.go:60]: Yea ideally I'd want to have a couple separate datasources that you can actually use within the config resource to avoid this thing getting too out of control, but haven't quite spent the time to look into it.  not sure if it makes too much sense for datasource of the firewall config

File: vercel/resource_firewall_config.go [hunk 6]
```diff
@@ -0,0 +1,878 @@
...
+func (r *firewallConfigResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state FirewallConfig
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	out, err := r.client.GetFirewallConfig(ctx, state.ProjectID.ValueString(), state.TeamID.ValueString())
+	if err != nil {
```
Comment (dglsparsons) [inline vercel/resource_firewall_config.go:802]: Do we need to handle 404s here?

File: vercel/resource_firewall_config.go [hunk 7]
```diff
@@ -0,0 +1,878 @@
...
+	diags := req.Plan.Get(ctx, &plan)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	conf := plan.toClient()
+
+	out, err := r.client.PutFirewallConfig(ctx, conf)
+	if err != nil {
+		diags.AddError("failed to create firewall config", err.Error())
```
Comment (dglsparsons) [inline vercel/resource_firewall_config.go:821]: missing `return`?

File: vercel/resource_firewall_config.go (hunk 6 above)
Comment (sueplex) [reply vercel/resource_firewall_config.go:802]: YES.  Actually ran into a couple bugs testing locally that made no sense.

File: go.mod [hunk 8]
```diff
@@ -3,8 +3,9 @@ module github.com/vercel/terraform-provider-vercel
 go 1.21
 
 require (
-	github.com/hashicorp/terraform-plugin-framework v1.7.0
-	github.com/hashicorp/terraform-plugin-go v0.22.1
+	github.com/hashicorp/terraform-plugin-framework v1.10.0
+	github.com/hashicorp/terraform-plugin-framework-validators v0.13.0
+	github.com/hashicorp/terraform-plugin-go v0.23.0
```
Comment (dglsparsons) [inline go.mod:8]: I usually try to do package bumps as a separate PR, but not worth separating out now.

**Review by craigandrews — APPROVED** [review]

**Review by dglsparsons — APPROVED** [review]

File: vercel/resource_firewall_config.go [hunk 9]
```diff
@@ -0,0 +1,991 @@
...
+}
+
+func (r *firewallConfigResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: `
+Define Custom Rules to shape the way your traffic is handled by the Vercel Edge Network.`,
+		Blocks: map[string]schema.Block{
+			"managed_rulesets": schema.SingleNestedBlock{
+				Description: "The managed rulesets that are enabled.",
+				Blocks: map[string]schema.Block{
```
Comment (jasonodonnell) [inline vercel/resource_firewall_config.go:46]: A helper function to populate these might make this more readable.
